File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/item/armor/effects/SpeedModifier.java
package generations.gg.generations.core.generationscore.common.world.item.armor.effects;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableMultimap;
import com.google.common.collect.Multimap;
import generations.gg.generations.core.generationscore.common.world.item.armor.CustomAttributeModifier;
import generations.gg.generations.core.generationscore.common.world.item.armor.GenerationsArmorItem;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.entity.ai.attributes.Attribute;
import net.minecraft.world.entity.ai.attributes.AttributeModifier;
import net.minecraft.world.entity.ai.attributes.Attributes;
import net.minecraft.world.item.ItemStack;

import java.util.UUID;

public class SpeedModifier implements CustomAttributeModifier {
    private final AttributeModifier elementalBootsModifier;
    private final AttributeModifier elementalBootsModifier2x;

    public SpeedModifier(float value) {
        this.elementalBootsModifier = new AttributeModifier(UUID.fromString("10ae6bcc-5b15-41b1-ba51-b6101e178401"), Attributes.MOVEMENT_SPEED.getDescriptionId(), value, AttributeModifier.Operation.MULTIPLY_BASE);
        this.elementalBootsModifier2x = new AttributeModifier(UUID.fromString("de4f0383-fcf9-4ba7-8ffc-0767c1ead7b9"), Attributes.MOVEMENT_SPEED.getDescriptionId(), value * 2.0f, AttributeModifier.Operation.MULTIPLY_BASE);
    }

    @Override
    public void getAttributeModifiers(ImmutableMultimap.Builder<Attribute, AttributeModifier> builder, EquipmentSlot equipmentSlot, ItemStack itemStack, GenerationsArmorItem generationsArmorItem) {
        if(equipmentSlot == generationsArmorItem.getType().getSlot()) builder.put(Attributes.MOVEMENT_SPEED, isDoubled(itemStack) ? this.elementalBootsModifier2x : this.elementalBootsModifier);
    }

    private boolean isDoubled(ItemStack itemStack) {
        CompoundTag armorEffectTagElement = itemStack.getTagElement("ArmorEffect");
        if (armorEffectTagElement == null) return false;
        return armorEffectTagElement.getBoolean("DoubleSpeed");
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/item/armor/effects/SpeedModifier.kt
package generations.gg.generations.core.generationscore.common.world.item.armor.effects

import generations.gg.generations.core.generationscore.common.GenerationsCore
import generations.gg.generations.core.generationscore.common.world.item.armor.ArmorTickEffect
import generations.gg.generations.core.generationscore.common.world.item.armor.GenerationsArmorItem
import net.minecraft.core.component.DataComponents
import net.minecraft.world.entity.EquipmentSlotGroup
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.ai.attributes.AttributeModifier
import net.minecraft.world.entity.ai.attributes.Attributes
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.component.ItemAttributeModifiers
import net.minecraft.world.level.Level


//TODO: Figure out how to get this working again.
class SpeedModifier(val value: Double, val valueFullSet: Double = value) : ArmorTickEffect {
    val baseModifier = GenerationsCore.id("armor_speed_boost");

    fun ItemAttributeModifiers.replaceModifierValue(fullSet: Boolean): ItemAttributeModifiers {
        val attribute = Attributes.MOVEMENT_SPEED

        val updated = this.modifiers()
            .filter { entry: ItemAttributeModifiers.Entry -> !(entry.matches(attribute, baseModifier) && entry.slot == EquipmentSlotGroup.LEGS) }
            .toMutableList()

        val replacement = AttributeModifier(baseModifier, if(fullSet) valueFullSet else value, AttributeModifier.Operation.ADD_MULTIPLIED_BASE)

        updated.add(ItemAttributeModifiers.Entry(attribute, replacement, EquipmentSlotGroup.LEGS))

        return ItemAttributeModifiers(updated, this.showInTooltip())
    }

    override fun inventoryTick(
        itemStack: ItemStack,
        world: Level,
        entity: LivingEntity,
        slotId: Int,
        isSelected: Boolean,
        generationsArmorItem: GenerationsArmorItem,
    ) {
        if(world.isClientSide()) return
        itemStack.update(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY, { it.replaceModifierValue(ArmorTickEffect.isWearingFullSet(entity, generationsArmorItem.material))})?: return
    }

//    override fun getAttributeModifiers(
//        builder: ImmutableMultimap.Builder<Attribute?, AttributeModifier>,
//        equipmentSlot: EquipmentSlot,
//        itemStack: ItemStack,
//        generationsArmorItem: GenerationsArmorItem?,
//    ) {


//        if(equipmentSlot == EquipmentSlot.LEGS && generationsArmorItem?.type?.slot == equipmentSlot) builder.put(Attributes.MOVEMENT_SPEED, itemStack ? this.elementalBootsModifier2x : this.elementalBootsModifier);
//    }

    //    private boolean isDoubled(ItemStack itemStack) {
    //        CompoundTag armorEffectTagElement = itemStack.getTagElement("ArmorEffect");
    //        if (armorEffectTagElement == null) return false;
    //        return armorEffectTagElement.getBoolean("DoubleSpeed");
    //
}
